### ISSUE TITLE
Automate testing of key user flows

### ISSUE THREAD

**Issue by leonstafford** [sleepbusorg/sleepbus.org#16]
_From @leonstafford on May 27, 2017 3:17_

### Background

We currently have no feedback of breakage and as we're actively developing and refactoring the existing codebase, plus wanting to get to continuous integration, this will be useful.

Preferably Python/Ruby/JS based to be a bit easier to work on.

#### Scenarios

 - [ ] User signup scenario
 - [ ] User sign-in scenario
 - [ ] One-time donation scenario
 - [ ] Monthly donation scenario 

#### Framework

 - [ ] Run in own Docker container
 - [ ] Use headless browser
 - [ ] Can run locally or in CI 

_Copied from original issue: sleepbusorg/sleepbus_org#11_

**Comment by bakersemail** [comment]
Hi Leon,

Looks like I can use my local development environment again. For this issue, I've found a couple candidates I would like to run by you:
1) Robot + Selenium python: https://bitbucket.org/robotframework/webdemo/overview
2) Straight Selenium python: http://selenium-python.readthedocs.io/getting-started.html#simple-usage

Selenium can be set up with a headless driver or even have a real browser which has it's display piped to a test display. The headless drivers are generally faster though.

My thoughts are to have a new or use the existing Docker file to set up the local development environment, attach to the running container and then start the tests. Though I wonder how we can get the test results published/viewable? Do you have a preference for this?

**Comment by leonstafford** [comment]
@bakersemail - we got a free year with browserstack, so can use them as our browser, with benefits of being able to run across a few different browser/os combos.

I'd like to see Robot in action, for no other reason than I've not used it before 😸 But I know it's well supported.

We've started writing "Gherkin" scenarios, so if we could plugin the Cucumber/Gherkin into the Robot/Selenium/Python mix, would be great. 

Containered would be preferred, just to keep local dev environment under a 1-script process for quick entry, but as the tests will run on CI anyway, it's not the end of the world if we need local python setup for tests, so long as it doesn't block the current dev flow for a new developer.

For publishing the tests, we'll have visibility via CircleCI and BrowserStack (it can record screencasts of the tests for us, even!). That should be enough for now, else what I've done in the past is used the G Doc Python libs and published results to shared docs and/or email. 

TL;DR - you had me at Python 🐍

**Comment by bakersemail** [comment]
Hi Leon,

I've created a sample project: https://github.com/bakersemail/test-robot-sample
Can you review and let me know if this is an okay approach to go with? It's just opening up Google's homepage and asserting the title so far.

**Comment by leonstafford** [comment]
@bakersemail - looks good! If you can add in the Given/When/Then syntax to examples, good to go:

https://blog.codecentric.de/en/2009/11/givenwhenthen-and-example-tables-using-the-robot-framework/

We have some prepared for sleepbus here:

https://docs.google.com/document/d/1nQU_1xPU5nKDtLxuJBXoSBdgZovVSo951zBZ_tGbZOg/edit

**Comment by leonstafford** [comment]
http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#ignoring-given-when-then-and-but-prefixes

**Comment by bakersemail** [comment]
Should be easy enough to add. I'll have a go at it this afternoon.

**Comment by bakersemail** [comment]
@leonstafford Hi Leon,
I've put together some given when then cases to that project of mine I linked. Let me know if this is how you'd like to progress and I'll try to piece together some basic scenarios for sleepbus.

Do you know where in the code base and how you like this triggered during builds (or outside of builds)?

**Comment by leonstafford** [comment]
Sweet, that looks perfect!

Please add it to the sleepbus.org project and I think 2 triggers for running:

 - once at end of `destroy_rebuild` script
 - other times by running another shell script, thus allowing easy triggering in commit hooks or in CircleCI build script

Will need to read some env vars from parent shell if possible, such as web address, login username, password, etc. Else we can add to existing .env file, like other app config is now

**Comment by bakersemail** [comment]
Hi @leonstafford 
I've created a branch with the tests that executes a signup. Unfortunately, I am quite inexperienced with Docker and don't know how to connect the test container with the web server in sleepbusweb. Perhaps you can lend you experience in this and have them connect?

So far I have only been able to run it against the web docker container and my host through running robot directly, rather than within the docker container.

See: https://github.com/sleepbusorg/sleepbus.org/tree/automate_testing_of_key_user_flows/test

**Comment by leonstafford** [comment]
@bakersemail sure thing. Docker will assign the container/image name as the hostname in our setup, so from another container, it should always be accessible as `sleepbusweb`. I'll assign this issue to myself and get it running today.

If you are able to look at https://github.com/sleepbusorg/sleepbus.org/issues/77 next or let me know if it's not suitable, I keep avoiding it myself 👻

**Comment by leonstafford** [comment]
container is able to see sleepbus web container via internal IP, is `172.17.0.4`, but codeigniter base url and redirect rules make it try to hit `www.172.17.0.4`. Thinking to pass in base url override and spin up a new instance of the sql and web containers for the test script

**Comment by leonstafford** [comment]
@bakersemail - got close the other day with getting it working with the container, but been caught up, should have a good stretch of time tomorrow. 

@mop13 has updated the whole app's scenarios here and I'll start coding them up soon as I get the container connectivity sorted. 

Web container can see sql container via hostname, but not the other way around. 

_Note to self, will next check Avahi/DNS discovery setup in the images / check port availability / capture web instance IP in wrapper script and pass to test container_

**Comment by leonstafford** [comment]
I now have this testing against my local dev containers, but these have specific hostnames, not localhost for the web container. Seeing if that can be also read from env file

**Comment by leonstafford** [comment]
Will look into reconfiguring the provisioning setup to use docker networking mode that shares host IP https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach

**Comment by leonstafford** [comment]
@bakersemail sorry for delay, had another session with this today. Looks like we've got a solution to the test container not seeing the other container easily. I'm going to rejig things yet again so that the test script spins up an isolated group of containers vs sharing the local dev ones. 

Will ping you again when ready to test 🚧
